src/bluesearch/database/article.py

```diff
@@ -184,17 +183,69 @@ def doi(self) -> str | None:
         """
         return None
 
+    @staticmethod
+    def get_uid_from_identifiers(identifiers: tuple[str | None, ...]) -> str:
+        """Generate a deterministic UID for a list of given paper identifiers.
+
+        Papers with the same values for the given identifiers get the same UID.
+
+        Missing values should have the value `None`, which is considered a value
+        by itself. Then, identifiers `(a, None)` and identifiers `(a, b)` have
+        two different UIDs.
+
+        Papers with all given identifiers unspecified have `None` as UID.
+
+        Parameters
+        ----------
+        identifiers
+            Values of the identifiers.
+
+        Returns
+        -------
+        str
+            A deterministic UID computed from the identifiers.
+
+        Raises
+        ------
+        ValueError
+            If all identifiers are `None`.
+        """
+        if all(x is None for x in identifiers):
+            raise ValueError(
+                f"Identifiers = {identifiers} are all `None`, UID cannot be computed."
+            )
+        else:
+            data = str(identifiers).encode()
+            hashed = hashlib.md5(data).hexdigest()  # nosec
+            return hashed
+
     @property
-    def uid(self) -> str | None:
-        """Generate unique ID of the article based on different identifiers.
+    def uid(self) -> str:
+        """Generate deterministic UID for an article.
+
+        The UID is usually created by hashing the identifiers of the article.
+        If no identifier is available, then the unique ID is computed by hashing
+        the whole content of the article.
 
         Returns
         -------
-        str or None
-            If at least one identifier exists, unique id is created. Otherwise,
-            the returned uid is None.
+        str
+            A deterministic UID.
         """
-        return generate_uid((self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi))
+        identifiers = (self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi)
```

Comment:
AFAIK `medrxiv` and `biorxiv` articles are guaranteed to have a DOI

Reply:
> Is it enough with those four identifiers ?

For the time being the answer should be Yes, because we currently work with papers from 5 sources and all of them seem to be covered:
- all `PubMed` papers have PubMed-ID
- all `PMC` papers have PMC-ID
- all `arXiv` papers have ArXiv ID
- all `bioRxiv` / `medRxiv` papers have DOI ([source](https://www.biorxiv.org/about/FAQ))

Now, this being said, it is totally possible that we are missing some cases, especially in the case of new sources that we may consider in the future.

At the same time, after this PR we will have a safety net for all cases where all the 4 identifiers are `None`.

Reply:
Perfect! Thank you both for the explanation :)